### PR TITLE
[SPARK-25832][SQL] remove newly added map related functions from FunctionRegistry

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

According to the discussion in RC4 voting thread and https://issues.apache.org/jira/browse/SPARK-25829, Spark current has a very weird behavior when we have duplicated keys in map. The newly added map related functions in 2.4 make it worse.

Before we entire fix the map behavior, we should not expose these functions to users

## How was this patch tested?

N/A

### CODE REVIEW

**Comment by cloud-fan** [comment]
cc @dongjoon-hyun @gatorsmile @rxin

**Review by gatorsmile — COMMENTED** [review]
LGTM pending Jenkins. 

BTW, you also need to remove the corresponding test cases that rely on function registry.

**Review by dongjoon-hyun — COMMENTED** [review]
@cloud-fan and @gatorsmile .

I agree that this is the quickest way. However, if you don't mind, I want to ask to remove those expressions(`MapFilter`) and test suites completely in `branch-2.4`.

According to this PR, it's too easy to expose back. Since these are very valuable functions, developers will use this hack definitely. Also, since we will decide how to change the behavior in Spark 3.0, it would be great if we completely prevent this exposure.
```scala
scala> spark.sessionState.functionRegistry.createOrReplaceTempFunction("map_filter", x => org.apache.spark.sql.catalyst.expressions.MapFilter(x(0),x(1)))
```

**Review by dongjoon-hyun — CHANGES_REQUESTED** [review]
I made a PR to you. Could you review and merge https://github.com/cloud-fan/spark/pull/11 ?

**Comment by dilipbiswal** [comment]
retest this please

**Comment by dongjoon-hyun** [comment]
@cloud-fan . That's sounds like a Tech. Preview for the advance users, doesn't it?

It looks like an excuse to ignore the whole context of the discussion and to try to ship in any way. I cann't agree with you with this approach. IMO, we need to fix this or we need to not ship any code like this to any users.
> @dongjoon-hyun if there are advanced users who know all the background, and still want to use these functions, why shall we stop them? If end users can't hit the bug with public APIs, I think we are fine.

IIUC, the reason why we are not fixing this is that we want to keep the release cadence. If then, please simply remove this. I already gave you the code.
- https://github.com/cloud-fan/spark/pull/11

**Comment by dongjoon-hyun** [comment]
I'm just confused here. Shall we finish the discussion on the email thread? @cloud-fan and @gatorsmile . If the decision is officially made like that (providing tech. preview to advance users) in the email thread, I'm okay with this.

**Comment by rxin** [comment]
We seem to be splitting hairs here. Why are we providing tech preview to
advanced users? Are you saying they construct expressions directly using
internal APIs? I doubt that’s tech preview.

Users can construct a lot of invalid plans that lead to weird semantics or
behaviors if they try, and this doesn’t really make it worse.

In either case it’s not that difficult to remove them and add them back so
I could see it going  either way.

**Comment by dongjoon-hyun** [comment]
Thank you, @rxin . In that case, +1 for complete removal.
It's easier for us to add the expressions back instead of updating exising expressions.

**Comment by dongjoon-hyun** [comment]
@cloud-fan . I'll update my PR to you once more.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala
Comment (dongjoon-hyun) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: To remove `map_entries`, we need to R together. I'll include my PR to you, @cloud-fan .
cc @felixcheung

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Also, we need to remove `map_entires` from `functions.scala`, too.

Comment (dongjoon-hyun) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/FunctionRegistry.scala]: Also, cc @HyukjinKwon . We need to remove `map_entries` in Python.

**Comment by gatorsmile** [comment]
I will submit a PR to revert all these changes. Thanks!

**Comment by felixcheung** [comment]
Please ping me on the R removal.

________________________________

**Pull request closed without merging**
